core/src/main/scala/neotypes/Session.scala

```diff
@@ -1,40 +1,45 @@
 package neotypes
 
-import neotypes.implicits.syntax.async._
-import neotypes.utils.stage._
+import internal.syntax.async._
+import internal.syntax.stage._
+
 import org.neo4j.driver.v1.{Session => NSession, Transaction => NTransaction}
 
 import scala.language.higherKinds
 
-final class Session[F[_]](session: NSession)(implicit F: Async[F]) {
-  def beginTransaction(): F[Transaction[F]] =
-    F.async { cb =>
-      session
-        .beginTransactionAsync()
-        .accept { tx: NTransaction => cb(Right(new Transaction(tx)(F))) }
-        .recover { ex: Throwable => cb(Left(ex)) }
-    }
+final class Session[F[_]](private val session: NSession) extends AnyVal {
+  def transaction[R[_]](implicit F: Async.Aux[F, R]): R[Transaction[F]] =
+    F.resource(createTransaction) { transaction => transaction.close }
 
-  def close(): F[Unit] =
+  private[neotypes] def createTransaction(implicit F: Async[F]): F[Transaction[F]] =
     F.async { cb =>
       session
-        .closeAsync()
-        .accept { _: Void => cb(Right(())) }
+        .beginTransactionAsync()
+        .accept { tx: NTransaction => cb(Right(new Transaction(tx))) }
         .recover { ex: Throwable => cb(Left(ex)) }
     }
 
-  def transact[T](txF: Transaction[F] => F[T]): F[T] =
-    beginTransaction().flatMap { t =>
+  def transact[T](txF: Transaction[F] => F[T])(implicit F: Async[F]): F[T] =
+    createTransaction.flatMap { t =>
       val result = for {
         res <- txF(t)
-        _ <- t.commit()
+        _ <- t.commit
       } yield res
 
-      result.recoverWith { case ex =>
-        for {
-          _ <- t.rollback()
-          res <- F.failed[T](ex)
-        } yield res
+      result.recoverWith {
+        case ex =>
+          for {
```

Comment:
I think this is way more idiomatic without for-comprehension (and it's also a one-liner):

```scala
t.rollback *> F.failed[T](ex)
```

Oh... And it's more performant if you're not using `better-monadic-for`.

Reply:
You are right. Sadly we do not have `*>`. But I may just write the `flatMap`.